### PR TITLE
[WIP] OJ-1960: Set up SonarCloud scanning on PRs and merges

### PR DESCRIPTION
**Pull request by CharlesIC**
_[This repo in SonarCloud](https://sonarcloud.io/project/overview?id=di-ipv-cri-toy-api)_

**Add SonarCloud config and workflow**

The reusable workflow gets coverage results from the calling workflow.
Add a standalone workflow to run SonarCloud on merges to main.
Add a Check PR workflow to run SonarCloud on PRs, alongside other checks.

**Add a workflow to verify PRs**

- Make unit tests and pre-commit reusable workflows
- Use the pre-commit action from the DI GitHub Actions library
- Add jest config to use when running in GitHub Actions - this uses a CI-friendly results reporter

Add one workflow to run all checks required for a PR to pass, and use reusable workflows, which allows to:
   - define dependencies (for example, SonarCloud needs unit tests to get the coverage reports)
   - make it easier to see all failures for a PR in the UI (only one workflow needs to be opened)

Screenshot 2023-10-10 at 14 49 53

**Update jest config and commands**

- Update jest config to correctly collect coverage
- Use default config values
- Update the scripts to make coverage a separate command
- Exclude files in the build directory from jest

---

- Add permissions and concurrency to all workflows
- Apply consisitent formatting to all workflows.

**BAU: Upgrade pre-commit hooks to the latest versions**

Add a pre-commit command to package.json, so pre-commit can be easily run on-demand.

Prettier v3.0 changed the default setting for the trailing commas option. Set it to the previous default in the config file to avoid having to update all the files (this can be done at a later time).
Reference: https://prettier.io/docs/en/options.html#trailing-commas

### CODE REVIEW

File: tests/favourite-handler.test.ts [hunk 1]
```diff
@@ -230,8 +230,7 @@ describe("favourite-handler.ts", () => {
   });
 
   it("should send the correct toy api response to txma", async () => {
-    const mockFetch = getMockFetch(404);
-    global.fetch = mockFetch;
+    global.fetch = getMockFetch(404);
```
Comment (merlinc) [inline tests/favourite-handler.test.ts]: This doesn't need to be done as part of this PR, and may not apply to the other repos

File: package.json [hunk 2]
```diff
@@ -4,13 +4,15 @@
   "description": "API for the toy credential issuer",
   "main": "index.js",
   "scripts": {
+    "pre-commit": "pre-commit run --all-files --show-diff-on-failure",
```
Comment (merlinc) [inline package.json]: Have you installed `pre-commit` and run `pre-commit install`.

I've noticed incidents where `pre-commit` doesn't run in a repo, and I'm not sure what causes it - especially given the tool is language independent, connected to Git lifecycles and shouldn't need extra tooling.

File: .prettierrc.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (merlinc) [inline .prettierrc.json]: DI has no recorded opinions on prettier, so let's stick with their defaults ... even if this means a big update as we move up to v3

File: .github/workflows/check-pr.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    uses: ./.github/workflows/sonarcloud.yml
+    secrets:
+      github-token: ${{ secrets.GITHUB_TOKEN }}
+      sonar-token: ${{ secrets.SONAR_TOKEN }}
+
+  pre-commit:
+    name: pre-commit
+    runs-on: ubuntu-latest
+    steps:
+      - name: Run pre-commit
```
Comment (merlinc) [inline .github/workflows/check-pr.yml]: What was the reasoning for creating our own GHA, as opposed to using the existing & functional but deprecated one [supplied by pre-commit](https://github.com/pre-commit/action)?

File: .github/workflows/unit-tests.yml [hunk 5]
```diff
@@ -1,24 +1,38 @@
...
-      - reopened
-      - ready_for_review
-      - synchronize
+name: Unit tests
+
+on: workflow_call
+permissions: {}
+
+concurrency:
+  group: unit-tests-${{ github.workflow }}-${{ github.head_ref || github.ref_name }}
```
Comment (merlinc) [inline .github/workflows/unit-tests.yml]: Is there any GHA documentation on how this concurrency / group block works?

File: .github/workflows/unit-tests.yml [hunk 6]
```diff
@@ -1,24 +1,38 @@
...
-      - name: Setup Node.js
+      - name: Pull repository
+        uses: actions/checkout@v4
+
+      - name: Set up Node
         uses: actions/setup-node@v3
         with:
-          node-version: 18.x
-      - name: Install Dependencies
+          cache: npm
```
Comment (merlinc) [inline .github/workflows/unit-tests.yml]: Removing the node version will leave it up to the [system installed version](https://github.com/actions/setup-node#usage).

If we don't want to be updating this all the time we should either use `node-version-file` and point it at an `.npmrc` file, or set it as `LTS` and have it automatically upgrade once a year.

File: .prettierrc.json (hunk 3 above)
Comment (merlinc) [reply .prettierrc.json]: @merlinc Actually, there's a lot going on in this PR, I'd also be happy to leave this out of this PR, and apply the change later.

File: package.json [hunk 7]
```diff
@@ -1,44 +1,46 @@
...
   },
-  "homepage": "https://github.com/alphagov/di-ipv-cri-toy-api#readme",
+  "homepage": "https://github.com/govuk-one-login/ipv-cri-toy-front/#readme",
   "devDependencies": {
     "@cucumber/cucumber": "8.6.0",
     "@types/aws-lambda": "^8.10.114",
     "@types/jest": "^29.5.0",
-    "@typescript-eslint/eslint-plugin": "5.10.2",
-    "@typescript-eslint/parser": "5.10.2",
+    "@typescript-eslint/eslint-plugin": "^6.7.5",
```
Comment (merlinc) [inline package.json]: Please install without the carets, as dependabot covers the ability to update to minor versions.

File: setEnvVars.js [hunk 8]
```diff
@@ -1,4 +1,4 @@
-process.env.AWS_STACK_NAME = "di-ipv-cri-toy-api";
+process.env.AWS_STACK_NAME = "ipv-cri-toy-api";
```
Comment (merlinc) [inline setEnvVars.js]: Something to pull out for another PR - but do we even need this file, or can we use [dotenv](https://www.npmjs.com/package/dotenv).

If we do need this file, let's add a comment at the top to stop me asking this question again in 3 months time.

**Review by merlinc — CHANGES_REQUESTED** [review]

File: package.json [hunk 9]
```diff
@@ -1,44 +1,46 @@
...
+  "name": "ipv-cri-toy-api",
   "description": "API for the toy credential issuer",
   "main": "index.js",
   "scripts": {
+    "pre-commit": "pre-commit run --all-files --show-diff-on-failure",
     "lint:prettier": "prettier --check lambdas",
     "lint:prettier-fix": "npx prettier --write .",
     "lint:eslint": "eslint .",
-    "lint": "npm run lint:prettier && npm run lint:eslint",
+    "lint": "npm run lint:eslint",
```
Comment (merlinc) [inline package.json]: :+1: We don't need prettier in this anymore, as it's covered by pre-commit.

File: package.json [hunk 10]
```diff
@@ -1,44 +1,46 @@
...
+    "lint:fix": "npm run lint:eslint -- --fix",
```
Comment (merlinc) [inline package.json]: Can we replace this with a [eslint pre-commit plugin](https://github.com/pre-commit/mirrors-eslint)?

**Review by merlinc — CHANGES_REQUESTED** [review]

File: .github/workflows/unit-tests.yml (hunk 5 above)
Comment (CharlesIC) [reply .github/workflows/unit-tests.yml]: https://docs.github.com/en/actions/using-jobs/using-concurrency

File: .github/workflows/check-pr.yml (hunk 4 above)
Comment (CharlesIC) [reply .github/workflows/check-pr.yml]: The mains reasons are that it's deprecated and using our own also allows us to reduce duplication since we can re-use one shared action in all the repos.

File: package.json (hunk 10 above)
Comment (CharlesIC) [reply package.json]: Added eslint plugin to pre-commit

File: .github/workflows/unit-tests.yml (hunk 6 above)
Comment (CharlesIC) [reply .github/workflows/unit-tests.yml]: Added back in
